### ISSUE TITLE
Research on DMPR-PS for keypoints detection

### ISSUE THREAD

**Issue by LamPeiLing** [Darcyddx/Active-Vision#5]
Reading <DMPR-PS: A NOVEL APPROACH FOR PARKING-SLOT DETECTION USING DIRECTIONAL MARKING-POINT REGRESSION> paper (1 hr)

**Comment by LamPeiLing** [comment]
Summary of the paper:

- proposed a novel CNN-based multi- attribute regression model to detect directional marking points. In our case, we might not  require to have the direction but we'll see how it's going on.
- two steps:

1. directional marking-point regression

-  used CNN-based multi-attribute regression model with regression that is designed for directional marking-points.
- this step is using DMPR of the model, catching all possible parking lots. 

2. parking-slot inference

- this step is filtering inappropriate marking-point pairs is necessary before the directional marking-point pairing. for example, some detected "slots" are too small for a car to park.
- it just done by simply calculating the distances between marking-points. 

Reflection: 
We could just using the model and pre-trained weight from first step and fine-tune using our own dataset. but this might require custom labelled dataset.

**Comment by LamPeiLing** [comment]
### Label dataset
if we want to use the label tool from DMPR-PS itself to label our tennis court dataset, something should be modified. In the research, there are two kinds of label:
1. marks

- the mark consists of a point, a circle, and an index number at the right bottom corner.
- we can just mark on the 14 points on the tennis court

2. slot

- after marking the points, on the right table, fill in the points that make a slot for a car, it stores in [first point, second point, type, angle] accordingly.
- we might need to remove this and simply link the points together for the tennis court. 

<img width="1464" alt="Image" src="https://github.com/user-attachments/assets/f18bfbb5-77f8-4446-8717-ab0e32d442a1" />

**Comment by LamPeiLing** [comment]
### Fine-tuning

- original output layer is 6 for parking-slot keypoints, we need to adjust it to 14 for tennis court
- uses darknet to extract feature
- need to modify data structures (in data/struct.py)
- can remove collect_threshold.py as we don't need that to determine different sizes of possible parking slots
- biggest change: preprocess dataset and those functions in label tool

Advantages of this approach:

- no need for heatmaps, use direct keypoint regression
- efficient keypoint refinement using DMPR-PS structure
- get fast inference where it outputs court keypoints directly

**Comment by LamPeiLing** [comment]
I decided to use other method
